### PR TITLE
Enumerable#each_cons support for accepting deque as a reuse buffer

### PR DESCRIPTION
**Pull request by yxhuvud**
This removes the restriction on `Enumerable#each_cons` and `Iterator#cons` to only accept arrays as a container to reuse. With this, it will be possible to use a Deque instead, for better performance.

Also adds some missing tests in iterator_spec.cr

Solves #7189

### CODE REVIEW

File: src/iterator.cr [hunk 1]
```diff
@@ -297,30 +297,26 @@ module Iterator(T)
...
   # * If *reuse* is an `Array`, this array will be reused
   # * If *reuse* is truthy, the method will create a new array and reuse it.
+  # * If *reuse* is falsey, the array will not be reused.
   #
   # This can be used to prevent many memory allocations when each slice of
   # interest is to be used in a read-only fashion.
   def cons(n : Int, reuse = false)
     raise ArgumentError.new "Invalid cons size: #{n}" if n <= 0
-    Cons(typeof(self), T, typeof(n)).new(self, n, reuse)
+    if reuse.is_a?(Bool)
```
Comment (bcardiff) [inline src/iterator.cr]: If reuse is nil this might be falling in the undesired branch. The docs state about falsey values.

Comment (yxhuvud) [reply src/iterator.cr]: Good point. Added a check for nil and a test for that case.

File: src/enumerable.cr [hunk 2]
```diff
@@ -281,14 +281,15 @@ module Enumerable(T)
   #
   # By default, a new array is created and yielded for each consecutive slice of elements.
   # * If *reuse* is given, the array can be reused
-  # * If *reuse* is an `Array`, this array will be reused
-  # * If *reuse* is truthy, the method will create a new array and reuse it.
+  # * If *reuse* is `true`, the method will create a new array and reuse it.
+  # * If *reuse* is a truthy, the passed `Array` or `Deque` will be used.
```
Comment (RX14) [inline src/enumerable.cr]: typo: a truthy

File: src/iterator.cr [hunk 3]
```diff
@@ -297,30 +297,26 @@ module Iterator(T)
...
-          @values = Array(T).new(@n)
-        end
-        @reuse = true
-      else
-        @values = Array(T).new(@n)
-        @reuse = false
-      end
+    def initialize(@iterator : I, @n : N, values : V, reuse)
+      @values = values
+      @reuse = reuse ? true : false
```
Comment (RX14) [inline src/iterator.cr]: Consider `@reuse = !!reuse`, it seems to be a fairly common idiom, are there objections?

**Comment by straight-shoota** [comment]
Should `reuse == true` use a `Deque` as default for better performance than array?
Or does it need the standard collection type?

**Comment by RX14** [comment]
@straight-shoota it'd break a lot of code, since a lot of code only takes arrays.

File: src/iterator.cr (hunk 3 above)
Comment (bcardiff) [reply src/iterator.cr]: No objections on my side, but we don't need to enforce it IMO.

Comment (yxhuvud) [reply src/iterator.cr]: I have no opinion on the matter, except that I didn't like spending 5 lines on it which is what the original code looked like after my changes in the logic.

**Comment by yxhuvud** [comment]
Should I rebase it down to three commits again?

**Comment by RX14** [comment]
yeah

File: src/iterator.cr [hunk 4]
```diff
@@ -297,30 +297,26 @@ module Iterator(T)
   # * If *reuse* is given, the array can be reused
   # * If *reuse* is an `Array`, this array will be reused
```
Comment (straight-shoota) [inline src/iterator.cr]: This should be updated to allow `Array`, `Deque` or any collection type with a matching interface.

File: spec/std/enumerable_spec.cr [hunk 5]
```diff
@@ -286,7 +286,16 @@ describe "Enumerable" do
       a.should be(reuse)
     end
 
-    it "returns running pairs with reuse = true" do
+    it "returns each_cons iterator with reuse = deque" do
+      reuse = Deque(Int32).new
+      iter = [1, 2, 3, 4, 5].each_cons(3, reuse: reuse)
+
+      a = iter.next
+      a.should eq(Deque(Int32).new([1, 2, 3]))
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/enumerable_spec.cr]: Any type with argless constructor and `#<<`. See https://crystal-lang.org/reference/syntax_and_semantics/literals/array.html#array-like-type-literal

File: spec/std/enumerable_spec.cr [hunk 6]
```diff
@@ -306,6 +315,16 @@ describe "Enumerable" do
       end
       array.should eq([[1, 2], [2, 3], [3, 4]])
     end
+
+    it "yields running pairs with reuse = deque" do
+      array = [] of Array(Int32)
+      reuse = Deque(Int32).new
+      [1, 2, 3, 4].each_cons(2, reuse: reuse) do |pair|
+        pair.should be(reuse)
+        array << pair.to_a
```
[the remark this answers is not in the document]
Comment (yxhuvud) [reply spec/std/enumerable_spec.cr]: Because Deque doesn't do structural equality. I tried the way you suggested at first, but `Deque(Foo).new != Deque(Foo).new`

[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/enumerable_spec.cr]: It should. See https://carc.in/#/r/5xoj

File: src/iterator.cr (hunk 4 above)
Comment (yxhuvud) [reply src/iterator.cr]: In practice, the only ones I found in stdlib that have `<<`, `shift` and `size` seems to be `Array` and `Deque`. I suppose this is where java or whatever would have reached for a Qeueable interface.

I'll synch the texts of the two methods, though.

File: spec/std/enumerable_spec.cr (hunk 6 above)
[the remark this answers is not in the document]
Comment (yxhuvud) [reply spec/std/enumerable_spec.cr]: Hmm. Changing it exposed a type issue. I'll try to figure out a way to fix that.

File: src/iterator.cr (hunk 4 above)
Comment (straight-shoota) [reply src/iterator.cr]: Yeah, the stdlib has only those two matching collections. But it should also be usable with any custom collection type.

File: spec/std/enumerable_spec.cr (hunk 6 above)
[the remark this answers is not in the document]
Comment (yxhuvud) [reply spec/std/enumerable_spec.cr]: Reused the solution from `Iterator`. Good enough?

File: src/enumerable.cr [hunk 7]
```diff
@@ -281,22 +281,22 @@ module Enumerable(T)
   #
   # By default, a new array is created and yielded for each consecutive slice of elements.
   # * If *reuse* is given, the array can be reused
-  # * If *reuse* is an `Array`, this array will be reused
-  # * If *reuse* is truthy, the method will create a new array and reuse it.
+  # * If *reuse* is `true`, the method will create a new array and reuse it.
+  # * If *reuse* is truthy, the passed `Array` or `Deque` will be used.
```
Comment (straight-shoota) [inline src/enumerable.cr]: An integer would also be truthy, but can't be passed.
I'd suggest something along:
```
If *reuse*  is an instance of `Array`, `Deque` or a similar collection type (implementing `#<<`, `#shift` and `#size`) it will be used.
```

We should probably have an abstract module for a generic list interface.

Comment (yxhuvud) [reply src/enumerable.cr]: Well, updated, and that list interface is way out of scope.

File: src/enumerable.cr [hunk 8]
```diff
@@ -283,22 +283,22 @@ module Enumerable(T)
...
   #
   # This can be used to prevent many memory allocations when each slice of
   # interest is to be used in a read-only fashion.
   def each_cons(count : Int, reuse = false)
-    if reuse
-      unless reuse.is_a?(Array)
-        reuse = Array(T).new(count)
-      end
-      cons = reuse
+    raise ArgumentError.new "Invalid cons size: #{count}" if count <= 0
```
Comment (straight-shoota) [inline src/enumerable.cr]: Please call it `Invalid count`, that's the name of the argument.

File: src/iterator.cr [hunk 9]
```diff
@@ -293,34 +293,30 @@ module Iterator(T)
...
   # * If *reuse* is given, the array can be reused
-  # * If *reuse* is an `Array`, this array will be reused
-  # * If *reuse* is truthy, the method will create a new array and reuse it.
+  # * If *reuse* is `true`, the method will create a new array and reuse it.
+  # * If *reuse*  is an instance of `Array`, `Deque` or a similar collection type (implementing `#<<`, `#shift` and `#size`) it will be used.
+  # * If *reuse* is falsey, the array will not be reused.
   #
   # This can be used to prevent many memory allocations when each slice of
   # interest is to be used in a read-only fashion.
   def cons(n : Int, reuse = false)
```
Comment (straight-shoota) [inline src/iterator.cr]: This argument should also be called `count` to line up with `Enumerable#each_cons`. But that would be a breaking change (and I don't think the other changes are breaking), so it can be left for another PR.

File: src/iterator.cr [hunk 10]
```diff
@@ -293,34 +293,30 @@ module Iterator(T)
...
+    if reuse.nil? || reuse.is_a?(Bool)
+      Cons(typeof(self), T, typeof(n), Array(T)).new(self, n, Array(T).new(n), reuse)
+    else
+      Cons(typeof(self), T, typeof(n), typeof(reuse)).new(self, n, reuse, reuse)
+    end
   end
 
-  private struct Cons(I, T, N)
+  private struct Cons(I, T, N, V)
     include Iterator(Array(T))
```
Comment (straight-shoota) [inline src/iterator.cr]: I think this should be changed to `Iterator(V)`?

File: src/iterator.cr (hunk 9 above)
Comment (Sija) [reply src/iterator.cr]: Pedantic: there's a double space in the comment above.

Comment (straight-shoota) [reply src/iterator.cr]: I noticed it, too, but didn't want to be too pedantic :P

File: src/iterator.cr (hunk 10 above)
Comment (yxhuvud) [reply src/iterator.cr]: Yes. But I wonder though, how to trigger the difference in a test.

File: src/iterator.cr (hunk 9 above)
Comment (straight-shoota) [reply src/iterator.cr]: @Sija, that would be breaking. Let's change it in another PR.

**Comment by yxhuvud** [comment]
(rebased)
